// cirruswave/deploy/docker/target.js
class TargetDeployment {
    constructor(destdir, tmpdir, tmpLaunchDir, tmpLaunchDir1, deployconfig, sysserviceconfigfilepath)
    {
        if(this.constructor.name  === "targetdeployment")
        {
            console.error("can not instantiate "+this.constructor.name);
            process.exit(1);
        }
        var methods = [];
        if(this.PrepCmdBefore === undefined)
            methods.push("PrepCmdBefore");
        if(this.PrepCmdAfter === undefined)
            methods.push("PrepCmdAfter");
        if(this.PrepCmdManager === undefined)
            methods.push("PrepCmdManager");
        if(this.PrepCmdWrite === undefined)
            methods.push("PrepCmdWrite");
        if(this.Servicify === undefined)
            methods.push("Servicify");
        if(methods.length)
        {
            console.error(methods+" required on "+this.constructor.name+" class")
            process.exit(1);
        }
        this.destdir = destdir;
        this.tmpdir = tmpdir;
        this.tmpLaunchDir = tmpLaunchDir;
        this.tmpLaunchDir1 = tmpLaunchDir1;
        this.deployconfig = deployconfig;
    }
}
module.exports = TargetDeployment;
